include/seqan3/alignment/matrix/alignment_trace_matrix.hpp

```diff
@@ -98,75 +99,148 @@ struct alignment_coordinate
 //!\endcond
-inline std::pair<std::list<gapped_database_t>, std::list<gapped_query_t>>
-alignment_trace(database_t && database, query_t && query, trace_matrix_t && matrix, alignment_coordinate const end_coordinate)
+inline std::pair<std::vector<gapped_database_alphabet_t>, std::vector<gapped_query_alphabet_t>>
+alignment_trace(database_t && database,
+                query_t && query,
+                trace_matrix_t && matrix,
+                alignment_coordinate const end_coordinate)
 {
+    using signed_size_t = std::make_signed_t<size_t>;
+
     constexpr auto N = trace_directions::none;
     constexpr auto D = trace_directions::diagonal;
     constexpr auto L = trace_directions::left;
     constexpr auto U = trace_directions::up;
-    size_t col = end_coordinate.seq1_pos + 1;
-    size_t row = end_coordinate.seq2_pos + 1;
+    signed_size_t col = end_coordinate.seq1_pos + 1;
+    signed_size_t row = end_coordinate.seq2_pos + 1;
 
-    assert(col <= database.size());
     assert(row <= query.size());
+    assert(col <= database.size());
+    assert(row < matrix.rows());
+    assert(col < matrix.cols());
 
-    std::list<gapped_database_t> gapped_database{};
-    std::list<gapped_query_t> gapped_query{};
+    std::deque<gapped_database_alphabet_t> gapped_database{};
+    std::deque<gapped_query_alphabet_t> gapped_query{};
 
     if (matrix.at(0, 0) != N)
-        throw std::runtime_error{"End trace must be NONE"};
+        throw std::logic_error{"End trace must be NONE"};
 
-    while(row >= 0 && col >= 0)
+    while(true)
     {
         trace_directions dir = matrix.at(row, col);
         if ((dir & L) == L)
         {
-            col = std::max<int>(col - 1, 0);
+            col = std::max<signed_size_t>(col - 1, 0);
             gapped_database.push_front(database[col]);
             gapped_query.push_front(gap::GAP);
         }
         else if ((dir & U) == U)
         {
-            row = std::max<int>(row - 1, 0);
+            row = std::max<signed_size_t>(row - 1, 0);
             gapped_database.push_front(gap::GAP);
             gapped_query.push_front(query[row]);
         }
         else if ((dir & D) == D)
         {
-            row = std::max<int>(row - 1, 0);
-            col = std::max<int>(col - 1, 0);
+            row = std::max<signed_size_t>(row - 1, 0);
+            col = std::max<signed_size_t>(col - 1, 0);
             gapped_database.push_front(database[col]);
             gapped_query.push_front(query[row]);
         }
         else
         {
             if (row == 0 || col == 0)
                 break;
-            throw std::runtime_error{"Trace not found"};
+            throw std::logic_error{"Trace not found"};
         }
+
     }
-    return {gapped_database, gapped_query};
+
+    return
+    {
+        {std::begin(gapped_database), std::end(gapped_database)},
+        {std::begin(gapped_query), std::end(gapped_query)}
+    };
```

Comment:
why this change?

Reply:
because the return and the local type does now differ.
So their is now a conversion from deque to vector.

Reply:
Wouldn't it be faster (and easier to read) to push_back() on a vector, call reverse() in the end and move it back out?

Reply:
I personally would not convert at all.

I doubt that this will be faster. And I'm not sure why it would be better to put something in reverse order just to reserve it again. From a data layout point it should be more expansive.

Reply:
Well now you are allocating for a deque. A deck is non-contiguous so you will likely allocate mulitple dispersed blocks of memory. Then on return you iterate over these blocks jumping around through memory and then allocating another time for the vector and copying the values. Since deque is non-contiguous you can never use memcpy for this.

Calling reverse on a vector seems a lot faster because you have a contiguous block and you swap the element internally...

Reply:
You know I had a vector and pushed it to the front, then @rrahn found it not cool. Then I changed it to a deque, because it has contiguous blocks, then @h-2 found it not cool. Whatever you say, I will not change this anymore for this pull request.

Reply:
The thing is that this function will change probably anyway, once we use different gapped sequence implementation.

Reply:
First, there is nothing to argue about performance before measuring performance. So this is likely to change since, for affine gaps this will change anyway.